gems/aws-sdk-s3/spec/presigner_spec.rb

```diff
@@ -162,14 +144,161 @@ module S3
           expect(url).to match(/^http:/)
         end
 
-        it 'uses the correct :endpoint scheme' do
+        it 'uses the configured :endpoint scheme' do
           client.config.endpoint = URI('http://example.com')
           signer = Presigner.new(client: client)
           url = signer.presigned_url(
             :get_object, bucket: 'aws-sdk', key: 'foo'
           )
           expect(url).to match(/^http:/)
         end
+
+        it 'supports virtual hosting' do
+          signer = Presigner.new(client: client)
+          url = signer.presigned_url(
+            :get_object, bucket: 'my.website.com', key: 'foo', virtual_host: true
+          )
+          expect(url).to match(/^https:\/\/my.website.com\/foo/)
+        end
+
+        it 'hoists x-amz-* headers to the query string' do
+          signer = Presigner.new(client: client)
+          url = signer.presigned_url(
+            :put_object, bucket: 'aws-sdk', key: 'foo', acl: 'public-read'
+          )
+          expect(url).to match(/x-amz-acl=public-read/)
+        end
+      end
+
+      describe '#presigned_request' do
```

Comment:
These tests are exactly the same as presigned_url except for the last. Shared examples is difficult here because of the return types.

Reply:
Yeah, it is a lot of duplication.  I think we get could shared examples working w/ the different return types using some sort of wrapper... but I dont think its really worth it here - as is, the tests are fairly readable.